crates/hc_service_check/src/main.rs

```diff
@@ -0,0 +1,363 @@
+                Cmd::PeerEvt(PeerConnectionEvent::State(PeerConnectionState::Connected)) => (),
+                Cmd::PeerEvt(PeerConnectionEvent::ICECandidate(mut ice)) => {
+                    if is_ice_relay(&mut ice) {
+                        t2o_snd.send(Cmd::Ice(ice)).unwrap();
+                    }
+                }
+                Cmd::PeerEvt(PeerConnectionEvent::DataChannel(data_chan, data_recv)) => {
+                    tokio::task::spawn(spawn_chan(
+                        data_chan,
+                        data_recv,
+                        start,
+                        chan_ready.clone(),
+                        rcv_done.clone(),
+                        msg_count,
+                    ));
+                }
+                Cmd::Offer(offer) => {
+                    c2.set_remote_description(offer).await.unwrap();
+                    let mut answer = c2.create_answer(AnswerConfig::default()).await.unwrap();
+                    println!(
+                        "created answer: {:?}",
+                        String::from_utf8_lossy(&answer.to_vec().unwrap())
+                    );
+                    c2.set_local_description(&mut answer).await.unwrap();
+                    t2o_snd.send(Cmd::Answer(answer)).unwrap();
+                    if let Some(ice_buf) = ice_buf.take() {
+                        for ice in ice_buf {
+                            c2.add_ice_candidate(ice).await.unwrap();
+                        }
+                    }
+                }
+                Cmd::Ice(ice) => {
+                    if let Some(ice_buf) = ice_buf.as_mut() {
+                        ice_buf.push(ice);
+                    } else {
+                        c2.add_ice_candidate(ice).await.unwrap();
+                    }
+                }
+                oth => panic!("unexpected: {oth:?}"),
+            }
+        }
+
+        Ok(())
+    })
+    .await
+    .map_err(Error::other)?
+}
+
+fn print_chan_ready_time(start: std::time::Instant) {
+    static CR: std::sync::Once = std::sync::Once::new();
+    CR.call_once(move || {
+        let elapsed = start.elapsed().as_secs_f64();
+        println!("\nchan ready in {elapsed} seconds");
+    });
+}
+
+fn print_rcv_done_time(start: std::time::Instant) {
+    static RD: std::sync::Once = std::sync::Once::new();
+    RD.call_once(move || {
+        let elapsed = start.elapsed().as_secs_f64();
+        println!("\nreceive done in {elapsed} seconds");
+    });
+}
+
+async fn spawn_peer(
+    config: PeerConnectionConfig,
+) -> (
+    PeerConnection,
+    tokio::sync::mpsc::UnboundedReceiver<PeerConnectionEvent>,
+) {
+    let (con, rcv) = PeerConnection::new(config).await.unwrap();
+    (con, rcv)
+}
+
+async fn spawn_chan(
+    data_chan: DataChannel,
+    mut data_recv: tokio::sync::mpsc::UnboundedReceiver<DataChannelEvent>,
+    start: std::time::Instant,
+    chan_ready: Arc<tokio::sync::Barrier>,
+    rcv_done: Arc<tokio::sync::Barrier>,
+    msg_count: usize,
+) {
```

Comment:
i just spotted the exit at the bottom.

```suggestion
) -> ! {
```

Reply:
I don't feel like the never type is canonical yet... we certainly don't use it in the codebase yet. I feel like we shouldn't bother with it until it becomes a standard enforceable clippy lint.